module/zfs/txg.c

```diff
@@ -549,8 +549,16 @@ txg_sync_thread(void *arg)
 			txg_thread_wait(tx, &cpr, &tx->tx_quiesce_done_cv, 0);
 		}
 
-		if (tx->tx_exiting)
-			txg_thread_exit(tx, &cpr, &tx->tx_sync_thread);
+		/*
+		 * Wait until the quiesce thread has exited to ensure every
+		 * quiesced txg has been synced before exiting.
+		 */
+		if (tx->tx_exiting) {
```

Comment:
Would it be sufficient to just change this to:

if (tx->tx_exiting && tx->tx_threads == 1) {
        txg_thread_exit(tx, &cpr, &tx->tx_sync_thread);
}

Reply:
@grwilson I am sorry about the long delay (a GitHub presence has become necessary for me only recently which reminded me of my short stint here ...).

Regarding your suggestion, I believe it would not be sufficient.

In the scenario given in the Motivation and Context, if `txg_quiesce_thread` (instead of `txg_sync_thread`) wakes up first and exits, then with your proposed edit, `txg_sync_thread` waking up next would still exit directly without syncing. 

This is why there is a test `if (tx->tx_quiesced_txg == 0)` before `txg_thread_exit` in my edit to allow the possible "one more sync before exiting".